File: core/src/promitech/colonization/gamelogic/combat/CombatController.java
package promitech.colonization.gamelogic.combat;

import promitech.colonization.GUIGameController;
import promitech.colonization.orders.move.MoveContext;
import promitech.colonization.ui.QuestionDialog;
import promitech.colonization.ui.QuestionDialog.OptionAction;

public class CombatController {
	
	private final GUIGameController guiGameController;
	private final CombatService combatService;
	
	private final Combat combat = new Combat();

	private final OptionAction<MoveContext> confirmWarDeclaration = new OptionAction<MoveContext>() {
		@Override
		public void executeAction(MoveContext payload) {
			showPreCombatDialog(payload);
		}
	};
	private final Runnable nextActiveUnitAction = new Runnable() {
	    @Override
	    public void run() {
	        guiGameController.nextActiveUnitAsGdxPostRunnable();
	    }
	};
    
	public CombatController(GUIGameController guiGameController, CombatService combatService) {
		this.guiGameController = guiGameController;
		this.combatService = combatService;
	}
	
	public void confirmCombat(MoveContext moveContext) {
		combat.init(moveContext.unit, moveContext.destTile);

		if (combat.canAttackWithoutConfirmation()) {
			showPreCombatDialog(moveContext);
		} else {
			CombatMsg combatMsg = new CombatMsg(combat);
			
			QuestionDialog questionDialog = new QuestionDialog();
			questionDialog.addQuestion(combatMsg.createAttackConfirmationMessageTemplate());
			questionDialog.addAnswer("model.diplomacy.attack.confirm", confirmWarDeclaration, moveContext);
			questionDialog.addOnlyCloseAnswer("cancel");
			guiGameController.showDialog(questionDialog);
		}
	}
	
	private void showPreCombatDialog(final MoveContext moveContext) {
		Runnable confirmCombatAction = new Runnable() {
			@Override
			public void run() {
                combatService.doConfirmedCombat(moveContext, combat, nextActiveUnitAction);
			}
		};
		SummaryDialog summaryDialog = new SummaryDialog(confirmCombatAction, combat);
		guiGameController.showDialog(summaryDialog);
	}
}

File: core/src/promitech/colonization/gamelogic/combat/CombatController.kt
package promitech.colonization.gamelogic.combat

import promitech.colonization.GUIGameController
import promitech.colonization.orders.move.MoveContext
import promitech.colonization.ui.QuestionDialog

class CombatController (
	private val guiGameController: GUIGameController,
	private val combatService: CombatService
) {
	
	private val combat: Combat = Combat()

    private val confirmWarDeclaration = QuestionDialog.OptionAction<MoveContext> {
        payload -> showPreCombatDialog(payload)
    }
    private val nextActiveUnitAction = Runnable {
        guiGameController.nextActiveUnitAsGdxPostRunnable()
    }

    fun confirmCombat(moveContext: MoveContext) {
		combat.init(moveContext.unit, moveContext.destTile)

        if (combat.canAttackWithoutConfirmation()) {
            showPreCombatDialog(moveContext)
        } else {
            val combatMsg = CombatMsg(combat)

            val questionDialog = QuestionDialog()
            questionDialog.addQuestion(combatMsg.createAttackConfirmationMessageTemplate())
            questionDialog.addAnswer("model.diplomacy.attack.confirm", confirmWarDeclaration, moveContext)
            questionDialog.addOnlyCloseAnswer("cancel")
            guiGameController.showDialog(questionDialog)
        }
    }

    private fun showPreCombatDialog(moveContext: MoveContext) {
        val confirmCombatAction = Runnable {
            combatService.doConfirmedCombat(moveContext, combat, nextActiveUnitAction)
        }
        val summaryDialog = SummaryDialog(confirmCombatAction, combat)
        guiGameController.showDialog(summaryDialog)
    }

}
